### PR TITLE
[412] feat(Positionable): implement "smart" repositioning logic

### PR DESCRIPTION
**Pull request by CITguy**
JIRA: SURF-1553

### LGTM's
- [x] Dev LGTM
- [x] Zoom LGTM

## NOTE
Position optimization logic is not perfect and breaks down in extreme edge cases, but it's a step in the right direction.  We'll be looking into correcting logic for these extreme edge cases in the near future.

Example: Scrolling Popovers on highly-constrained screen widths end up positioned off screen or with inconsistent dimensions.

### CODE REVIEW

**Comment by HelixUI** [comment]
Deploy preview for *helix-ui* processing.

Built with commit 36e4a789486b59c0106a313a294956a84417edd1

https://app.netlify.com/sites/helix-ui/deploys/5c46933c2cf6790008bef0cb

File: src/helix-ui/mixins/Positionable.js [hunk 1]
```diff
@@ -130,65 +211,106 @@ export const Positionable = (superclass) => {
...
         }
 
         /**
-         * Calculate fixed {x,y} coordinates relative to another HTML element.
+         * Calculate optimum position and fixed {x,y} coordinates needed
+         * to arrange the positionable element in relation to its
+         * `relativeElement`.
          *
-         * @returns {Object} Coordinate object with `x` and `y` properties.
+         * @returns {Object} optimum coordinate metadata
```
Comment (CITguy) [inline src/helix-ui/mixins/Positionable.js]: NOTE TO SELF: define `@typedef` for metadata object.

File: src/helix-ui/mixins/Positionable.js [hunk 2]
```diff
@@ -213,13 +335,89 @@ export const Positionable = (superclass) => {
...
         }
 
+        /**
+         * Calculate offset based on configuration and position
+         */
+        __getOffset () {
+            let offset = 0;
+            /*
+             * Remove offset if positioned on major axis
+             * so that the point of the arrow always aligns
```
Comment (CITguy) [inline src/helix-ui/mixins/Positionable.js]: NOTE TO SELF: `...point of _any_ arrow...`

File: src/helix-ui/elements/HXTooltipElement.js [hunk 3]
```diff
@@ -97,6 +96,16 @@ export class HXTooltipElement extends _ProtoClass {
         }
     }
 
+    /**
+     * @type {Integer}
+     */
+    set delay (value) {
```
[the remark this answers is not in the document]
Comment (CITguy) [reply src/helix-ui/elements/HXTooltipElement.js]: I'm gonna revert the delay prop for now. Will revisit in future.

File: src/helix-ui/mixins/Positionable.js [hunk 4]
```diff
@@ -125,70 +199,112 @@ export const Positionable = (superclass) => {
...
 
             let posRect = this.getBoundingClientRect();
             let relRect = this.relativeElement.getBoundingClientRect();
-            let calculate = offsetFunctionMap[this.position];
-            let opts = this.__getDeltas();
 
-            return calculate(posRect, relRect, opts);
+            let position = this.position;
+            let deltas = this.__getDeltas(position);
+            let calculate = offsetFunctionMap[position];
```
[the remark this answers is not in the document]
Comment (CITguy) [reply src/helix-ui/mixins/Positionable.js]: good to go. no changes needed

**Pull request merged**